### PR TITLE
[Processing] Performance improvement of Join Attributes by Location

### PR DESCRIPTION
**Pull request by phborba**
## Description
For large datasets (27,5k features to be joined with 30,5k features) algorithm did not run (the machine I was testing it on had 16Gb of ram and 2Gb of swap which were fully used and then crashed), so I have made the following changes to improve performance:

- Added spatial index usage on algorithm;
- Added a dictionary with all features of the index. This reduced the number of getFeatures calls, which has boosted the performance as well;
- Usage of python map to speedup large loops.

With these changes I could run the spatial join on the large dataset I was testing. With smaller ones, I could see that the algorithm was around 5 times faster.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by phborba** [comment]
@elpaso hi man how are you doing? Just a python for dummies question: I have just noticed that the pre-commit script has changed the lambda function I had defined inside my code into a def, do you know why?

**Comment by elpaso** [comment]
Sorry: no idea

**Comment by phborba** [comment]
@nyalldawson the check has failed due to timeout, what should I do?

**Comment by nyalldawson** [comment]
You can close and reopen to trigger a new build

File: python/plugins/processing/tests/Testing/Temporary/CTestCostData.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+---
```
Comment (luipir) [inline python/plugins/processing/tests/Testing/Temporary/CTestCostData.txt]: @phborba  is this file necessary?

**Comment by phborba** [comment]
Omg I did not see that I have committed that!
I’ll delete that file

**Comment by nyalldawson** [comment]
@phborba 

While this will speed up some cases, it's going to cause some large regressions in performance for other cases. Specifically - if a large point layer (say address points covering a country) is intersected with an individual locality polygon, it's now going to have to iterate over all the address point features up front in order to build the spatial index, vs the current approach which only ever fetches the points intersecting the polygon's bounding box from the provider. 

What we need here is some logic to decide when to apply the "fetch everything up front" approach as opposed to the current logic. I think this could potentially be done by comparing bounding boxes for the inputs -- if the two input's bounding boxes are similar, then the "fetch everything up front" approach used here should be used. But if the bounding boxes differ greatly in size then the current logic will be much faster and should be used instead.

**Comment by phborba** [comment]
I agree with you, I’ll try to figure something out. I was wondering of some sort of bounding box filter upon spatial index building, what do you think?

**Comment by nyalldawson** [comment]
> I agree with you, I’ll try to figure something out. I was wondering of some sort of bounding box filter upon spatial index building, what do you think?

Well, the logic switch needs to happen before any index is ever built (e.g. in the current optimised case we rely on on indices present on the provider's backend).

I think there's a few cases we need to handle:

1. (current optimised case) Selecting a small subset from a large layer A, based on a limited number of related features from layer B. E.g. "select address points from whole country address points layer, which fall within a set of counties". Can detect by checking extent of input datasets, if extent of layer A >> extent of features from layer B (or maybe number of features in layer A >> number of features in layer B?).

2. Selecting a large subset from a large layer A, based on a limited number of related features from layer B. E.g. "select address points from whole country address points layer, which fall within all but a few counties from the country". Can detect by checking extent of input datasets, if extent of layer A ~= extent of features from layer B (or counts of features?).

3. Selecting a subset from a layer A based on many features from layer B. E.g. "find the county from a counties layer where these 1 million address points are located within". Can detect by checking if total number of features from layer A << total number of features from layer B.

4. Selecting a large number of features from a large layer A when compared with a large layer B. E.g. "find me all address points from the whole country which fall have an address point". Could detect by checking if total features from layer A ~= total features from layer B and layers have a similar extent.

This discussion might best be moved to a QEP for wider feedback -- could you create an issue on https://github.com/qgis/QGIS-Enhancement-Proposals/issues so that suggestions can be gathered from the wider community before implementing any of this?
